crypto/x509/by_dir.c

```diff
@@ -257,165 +258,180 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
+    if (l == 0) {
+        hash[0] = 0;
+    }
+#ifndef OPENSSL_NO_MD5    
+    hash[1] = X509_NAME_hash_old(name); /* MD5 name hash */
+#else
+    hash[1] = 0;
+#endif
+
+    for (l = 0; l < 2; ++l) {
+        h = hash[l];
+        if (h == 0) {
+            continue; /* skip invalid hash */
         }
-        if (type == X509_LU_CRL && ent->hashes) {
-            htmp.hash = h;
-            if (!CRYPTO_THREAD_read_lock(ctx->lock))
+
+        for (i = 0; i < sk_BY_DIR_ENTRY_num(ctx->dirs); i++) {
+            BY_DIR_ENTRY *ent;
+            int idx;
+            BY_DIR_HASH htmp, *hent;
+
+            ent = sk_BY_DIR_ENTRY_value(ctx->dirs, i);
+            j = strlen(ent->dir) + 1 + 8 + 6 + 1 + 1;
+            if (!BUF_MEM_grow(b, j)) {
+                ERR_raise(ERR_LIB_X509, ERR_R_BUF_LIB);
                 goto finish;
-            idx = sk_BY_DIR_HASH_find(ent->hashes, &htmp);
-            if (idx >= 0) {
-                hent = sk_BY_DIR_HASH_value(ent->hashes, idx);
-                k = hent->suffix;
-            } else {
-                hent = NULL;
-                k = 0;
             }
-            CRYPTO_THREAD_unlock(ctx->lock);
-        } else {
-            k = 0;
-            hent = NULL;
-        }
-        for (;;) {
-            char c = '/';
-
-#ifdef OPENSSL_SYS_VMS
-            c = ent->dir[strlen(ent->dir) - 1];
-            if (c != ':' && c != '>' && c != ']') {
-                /*
-                 * If no separator is present, we assume the directory
-                 * specifier is a logical name, and add a colon.  We really
-                 * should use better VMS routines for merging things like
-                 * this, but this will do for now... -- Richard Levitte
-                 */
-                c = ':';
+            if (type == X509_LU_CRL && ent->hashes) {
+                htmp.hash = h;
+                if (!CRYPTO_THREAD_read_lock(ctx->lock))
+                    goto finish;
+                idx = sk_BY_DIR_HASH_find(ent->hashes, &htmp);
+                if (idx >= 0) {
+                    hent = sk_BY_DIR_HASH_value(ent->hashes, idx);
+                    k = hent->suffix;
+                } else {
+                    hent = NULL;
+                    k = 0;
+                }
+                CRYPTO_THREAD_unlock(ctx->lock);
             } else {
-                c = '\0';
+                k = 0;
+                hent = NULL;
             }
+            for (;;) {
+                char c = '/';
+
+    #ifdef OPENSSL_SYS_VMS
+                c = ent->dir[strlen(ent->dir) - 1];
+                if (c != ':' && c != '>' && c != ']') {
+                    /*
+                    * If no separator is present, we assume the directory
+                    * specifier is a logical name, and add a colon.  We really
+                    * should use better VMS routines for merging things like
+                    * this, but this will do for now... -- Richard Levitte
+                    */
```

Comment:
Nit: indentation is off by one character here.

Reply:
ef74616d2ea2cafcfe6c9cb04e00b69b0f6e8c6c